File: restcli-deno/src/executor.ts
import { type HttpRequest, applyVariables } from "./parser.ts";

export interface RequestResult {
  status: number;
  statusText: string;
  headers: Record<string, string>;
  body: string;
  duration: number;
  requestSize: number; // Size of request body in bytes
  responseSize: number; // Size of response body in bytes
  error?: string;
}

export class RequestExecutor {
  private baseUrl: string;

  constructor(baseUrl: string = "") {
    this.baseUrl = baseUrl;
  }

  async execute(
    request: HttpRequest,
    variables: Record<string, string>,
    profileHeaders: Record<string, string> = {}
  ): Promise<RequestResult> {
    const start = performance.now();

    try {
      // Apply variables
      const substituted = applyVariables(request, variables);

      // Build full URL
      let url = substituted.url;
      if (this.baseUrl && !url.startsWith("http")) {
        url = this.baseUrl + url;
      }

      // Merge headers: request headers override profile headers
      const headers = { ...profileHeaders, ...substituted.headers };

      // Make request
      const response = await fetch(url, {
        method: substituted.method,
        headers,
        body: substituted.body,
      });

      const duration = performance.now() - start;

      // Get response headers
      const responseHeaders: Record<string, string> = {};
      response.headers.forEach((value, key) => {
        responseHeaders[key] = value;
      });

      // Get body
      const body = await response.text();

      // Calculate sizes
      const requestSize = substituted.body ? new TextEncoder().encode(substituted.body).length : 0;
      const responseSize = new TextEncoder().encode(body).length;

      return {
        status: response.status,
        statusText: response.statusText,
        headers: responseHeaders,
        body,
        duration,
        requestSize,
        responseSize,
      };
    } catch (error) {
      const duration = performance.now() - start;
      return {
        status: 0,
        statusText: "Error",
        headers: {},
        body: "",
        duration,
        requestSize: 0,
        responseSize: 0,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

File: restcli-go/internal/executor/executor.go
package executor

import (
	"bytes"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/studiowebux/restcli/internal/types"
)

// Execute performs an HTTP request and returns the result
func Execute(req *types.HttpRequest) (*types.RequestResult, error) {
	startTime := time.Now()

	// Create HTTP request
	var bodyReader io.Reader
	requestSize := 0
	if req.Body != "" {
		bodyReader = bytes.NewBufferString(req.Body)
		requestSize = len(req.Body)
	}

	httpReq, err := http.NewRequest(req.Method, req.URL, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Set headers
	for key, value := range req.Headers {
		httpReq.Header.Set(key, value)
	}

	// Execute request
	client := &http.Client{
		Timeout: 30 * time.Second,
	}

	resp, err := client.Do(httpReq)
	duration := time.Since(startTime).Milliseconds()

	if err != nil {
		return &types.RequestResult{
			Error:       err.Error(),
			Duration:    duration,
			RequestSize: requestSize,
		}, nil
	}
	defer resp.Body.Close()

	// Read response body
	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return &types.RequestResult{
			Status:      resp.StatusCode,
			StatusText:  resp.Status,
			Error:       fmt.Sprintf("failed to read response body: %v", err),
			Duration:    duration,
			RequestSize: requestSize,
		}, nil
	}

	// Build response headers map
	headers := make(map[string]string)
	for key, values := range resp.Header {
		headers[key] = strings.Join(values, ", ")
	}

	result := &types.RequestResult{
		Status:       resp.StatusCode,
		StatusText:   resp.Status,
		Headers:      headers,
		Body:         string(bodyBytes),
		Duration:     duration,
		RequestSize:  requestSize,
		ResponseSize: len(bodyBytes),
	}

	return result, nil
}

// FormatDuration formats duration in milliseconds to human-readable string
func FormatDuration(ms int64) string {
	if ms < 1000 {
		return fmt.Sprintf("%dms", ms)
	}
	seconds := float64(ms) / 1000.0
	return fmt.Sprintf("%.2fs", seconds)
}

// FormatSize formats byte size to human-readable string
func FormatSize(bytes int) string {
	if bytes < 1024 {
		return fmt.Sprintf("%dB", bytes)
	}
	if bytes < 1024*1024 {
		return fmt.Sprintf("%.2fKB", float64(bytes)/1024.0)
	}
	return fmt.Sprintf("%.2fMB", float64(bytes)/(1024.0*1024.0))
}

// IsSuccessStatus returns true if status code is 2xx
func IsSuccessStatus(status int) bool {
	return status >= 200 && status < 300
}

// IsClientErrorStatus returns true if status code is 4xx
func IsClientErrorStatus(status int) bool {
	return status >= 400 && status < 500
}

// IsServerErrorStatus returns true if status code is 5xx
func IsServerErrorStatus(status int) bool {
	return status >= 500 && status < 600
}
